File: algorithms/algorithms-java/src/main/java/com/lanceolata/coding/algorithms/java/leetcode/Question_0524_Longest_Word_in_Dictionary_through_Deleting.java
package com.lanceolata.coding.algorithms.java.leetcode;

import java.util.List;

public class Question_0524_Longest_Word_in_Dictionary_through_Deleting {
    public String findLongestWord(String s, List<String> dictionary) {
        if (s == null || dictionary == null || dictionary.isEmpty()) {
            return null;
        }
        String res = "";
        for (String str : dictionary) {
            int i = 0;
            for (int j = 0; i < str.length() && j < s.length(); j++) {
                if (str.charAt(i) == s.charAt(j)) {
                    i++;
                }
            }
            if (i == str.length() && str.length() >= res.length()) {
                if (str.length() > res.length() || str.compareTo(res) < 0) {
                    res = str;
                }
            }
        }
        return res;
    }
}

File: algorithms/algorithms-cplusplus/leetcode/Question_0524_Longest_Word_in_Dictionary_through_Deleting.cc
#include <string>
#include <vector>

using namespace std;

class Solution {
 public:
  string findLongestWord(string s, vector<string>& dictionary) {
    string res = "";
    for (string str : dictionary) {
      int i = 0;
      for (char c : s) {
        if (i < str.size() && c == str[i]) {
          ++i;
        }
      }
      if (i == str.size() && str.size() >= res.size()) {
        if (str.size() > res.size() || str < res) {
          res = str;
        }
      }
    }
    return res;
  }
};
